pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,771 @@
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) updatePacketCapture(oldObj, newObj interface{}) {
+	newPc := newObj.(*crdv1alpha1.PacketCapture)
+	oldPc := oldObj.(*crdv1alpha1.PacketCapture)
+	if newPc.Generation != oldPc.Generation {
+		klog.InfoS("Processing PacketCapture UPDATE event", "name", newPc.Name)
+		c.enqueuePacketCapture(newPc)
+	}
+}
+
+func (c *Controller) deletePacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture DELETE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func nameToPath(name string) string {
+	return filepath.Join(packetDirectory, name+".pcapng")
+}
+
+func (c *Controller) worker() {
+	for c.processPacketCaptureItem() {
+	}
+}
+
+func (c *Controller) processPacketCaptureItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	if err := c.syncPacketCapture(key); err == nil {
+		c.queue.Forget(key)
+	} else {
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing PacketCapture, requeueing", "key", key)
+	}
+	return true
+}
+
+func (c *Controller) syncPacketCapture(pcName string) error {
+	cleanupStatus := func() {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		state := c.captures[pcName]
+		if state != nil {
+			if state.cancel != nil {
+				state.cancel()
+			}
+			delete(c.captures, pcName)
+		}
+	}
+
+	pc, err := c.packetCaptureLister.Get(pcName)
+	if apierrors.IsNotFound(err) {
+		c.cleanupPacketCapture(pcName)
+		cleanupStatus()
+		return nil
+	}
+
+	// Capture will not occur on this Node if a corresponding Pod interface is not found.
+	device := c.getTargetCaptureDevice(pc)
+	if device == nil {
+		return nil
+	}
+
+	if err := c.validatePacketCapture(&pc.Spec); err != nil {
+		klog.ErrorS(err, "Invalid PacketCapture", "name", pc.Name)
+		if updateErr := c.updateStatus(pc.Name, &packetCaptureState{err: err}); updateErr != nil {
+			klog.ErrorS(err, "Failed to update PacketCapture status", "name", pc.Name)
+		}
+		cleanupStatus()
+		return nil
+	}
+
+	state := func() *packetCaptureState {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		state := c.captures[pcName]
+		if state == nil {
+			state = &packetCaptureState{}
+			c.captures[pcName] = state
+		}
+		phase := state.phase
+		klog.InfoS("Syncing PacketCapture", "name", pcName, "phase", phase)
+		if phase != packetCapturePhasePending {
+			return state
+		}
+
+		if c.numRunningCaptures >= maxConcurrentCaptures {
+			err = fmt.Errorf("PacketCapture running count reach limit")
+		} else {
+			timeout := defaultTimeoutDuration
+			if pc.Spec.Timeout != nil {
+				timeout = time.Duration(*pc.Spec.Timeout) * time.Second
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			state.cancel = cancel
+			if err = c.startPacketCapture(ctx, pc, device); err != nil {
+				phase = packetCapturePhaseCompleted
+			} else {
+				phase = packetCapturePhaseRunning
```

Comment:
I might miss this when suggesting code, it should `c.numRunningCaptures += 1`.
I would expect an unit test to catch this: 
```go
func TestMultiplePacketCaptures(t *testing.T) {
        // Initialize the controller with > 16 captures
        // All captures should complete eventually with expected status
        // numRunningCaptures should be 0

        // Deleting all captures
        // `captures` should be empty
        // file directory should be empty
}
```

Reply:
added.